solutions/cf-920E.mdx

```diff
@@ -0,0 +1,233 @@
+---
+id: cf-920E
+source: CF
+title: Connected Components?
+author: Daniel Qiu
+---
+
+
+[Official Editorial](https://codeforces.com/blog/entry/57516)
+
+
+## Explanation
+## Solution 1 - Faster DFS
+
+Doing a standard DFS would be too slow. To speed it up, we have to do two
+optimizations.
+
+### Optimization 1
+
+You can store whether there is no edge between two nodes using an array of hashsets, instead of a adjacency list. If the input tells you that nodes $a$ and $b$ do node have an edge between them, you would insert node $a$ at $\texttt{array}[b]$, and node $b$ and $\texttt{array}[a]$ 
+This way you can find if there is no edge between two nodes in
+$\mathcal{O}(1)$ time. Using an adjacency matrix would take too much
+memory.
+
+### Optimization 2
+
+Rather than using a standard boolean array to check if a node is not visited, you
+can use a set of unvisited nodes. This way you only loop through the nodes that
+have not been visited yet and you can find the next unvisited node in $\mathcal{O}(\log N)$. You need to use ```upper_bound``` to find
+the next unvisited node after a DFS - an iterator may be invalidated due to the
+node it is pointing to being destroyed when you remove it from the unvisited set.
+
+### Proof
+
+We know that a node can either be skipped or visited. We also know that once a
+node is visited it cannot be revisited. Therefore you will always visit $N$ nodes once. We also know that the sum of skips can be at most $2M$ (each
```

Comment:
By skipped do you mean that it will never be visited, or that it will be visited in the future?

Reply:
If it means that it will never be visited, change it to "Therefore, you will visit at most $N$ nodes"

Reply:
It will be visited in the future.

Reply:
I think skipped might be a bit misleading. Just say something like we will visit all nodes once, and since we keep track of which nodes have been visited, we won't visit any node more than once.

Reply:
I think it is important to prove that the number of times you loop is not greater than 2m+n, otherwise, the solution could TLE